pkg/virt-handler/backendstorage/pod.go

```diff
@@ -0,0 +1,181 @@
+// newly created subdirectory. For example:
+// "/var/lib/libvirt/qemu/nvram -> /var/lib/libvirt/vm-state/nvram".
+func prepareVMStateDirectories(vmi *v1.VirtualMachineInstance, vmStateDir *safepath.Path) error {
+	if err := configureNonRootOwnerAndSelinuxContext(vmi, vmStateDir); err != nil {
+		return err
+	}
+
+	rootPath := unsafepath.UnsafeRoot(vmStateDir.Raw())
+	dirs := make(map[string]string)
+	if backendstorage.HasPersistentEFI(&vmi.Spec) {
+		dirs[services.PathForNVram(vmi)] = "nvram"
+	}
+	if backendstorage.HasPersistentTPMDevice(&vmi.Spec) {
+		dirs[services.PathForSwtpm(vmi)] = "swtpm"
+		dirs[services.PathForSwtpmLocalca(vmi)] = "swtpm-localca"
+	}
+	for srcRel, subDir := range dirs {
+		srcAbs := path.Join(rootPath, srcRel)
+		dstRel := path.Join(unsafepath.UnsafeRelative(vmStateDir.Raw()), subDir)
+		dstAbs := path.Join(unsafepath.UnsafeAbsolute(vmStateDir.Raw()), subDir)
+		createSymlink := true
+		if f, err := os.Lstat(srcAbs); err != nil && !errors.Is(err, os.ErrNotExist) {
+			return err
+		} else if err == nil { // if the file already exists
+			if f.Mode()&os.ModeDir != 0 {
+				if err := os.Remove(srcAbs); err != nil {
+					return err
+				}
+			} else if f.Mode()&os.ModeSymlink != 0 {
+				t, err := os.Readlink(srcAbs)
+				if err != nil {
+					return err
+				}
+				if t != dstRel {
+					return fmt.Errorf("symlink %s exists but pointed to an unexpected target %s", srcAbs, t)
+				}
+				createSymlink = false
+			} else {
+				return fmt.Errorf("unknown file type: %s, %d", srcAbs, f.Mode())
+			}
+		}
+
+		if err := os.MkdirAll(path.Dir(srcAbs), os.ModePerm); err != nil {
+			return err
+		}
+		if err := os.MkdirAll(dstAbs, os.ModePerm); err != nil {
+			return err
+		}
+		dst, err := safepath.JoinAndResolveWithRelativeRoot(rootPath, dstRel)
+		if err != nil {
+			return err
+		}
+		if configureNonRootOwnerAndSelinuxContext(vmi, dst); err != nil {
+			return err
+		}
+
+		if createSymlink {
+			log.DefaultLogger().V(2).Infof("Creating the symlink %s pointing to %s", srcAbs, dstRel)
+			if err := os.Symlink(dstRel, srcAbs); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// configureSwtpmDirsOwnership changes the owner and the group of the swtpm
+// directories to "tss" for a root VM in the filesystem backend storage case.
+// The block backend storage case is handled separately when the symlinks are
+// created, as the directories are different.
+//
+// In the root VM case, the swtpm process is started with user "tss" which
+// is a non-root user. Therefore, we give this user the access of the swtpm
+// directories by changing the owner and group to "tss".
+func configureSwtpmDirsOwnership(vmi *v1.VirtualMachineInstance, root *safepath.Path) error {
+	if !backendstorage.HasPersistentTPMDevice(&vmi.Spec) {
+		return nil
+	}
+	var uid, gid int
+	if utils.IsNonRootVMI(vmi) {
+		uid, gid = qemuUserID, qemuGroupID
+	} else {
+		uid, gid = tssUserID, tssGroupID
```

Comment:
Nit: consider adding a UID to util for TSS

Reply:
Could you elaborate?

Reply:
If you accept my suggestion above for the NonRootUID, then it would make sense to move the TSS const(s) to pkg/util/util as well. Feel free to ignore though, really not important.